### ISSUE TITLE
Getting NaN losses after a point

### ISSUE THREAD

**Issue by simplyrin2** [SamsungSAILMontreal/TinyRecursiveModels#28]

**Comment by bumi001** [comment]
I went through the code more carefully. Unrelated to the nan issue, I made the following change to q_halt_accuracy in models/losses.py.
```
                "q_halt_accuracy": (valid_metrics & (outputs["q_halt_logits"] >= 0) & ((outputs["q_halt_logits"] >= 0) == seq_is_correct)).sum(),
```
Without the above change we get incorrect numbers because the last equality is True when False == False.

Now back to the nan issue I am facing when it = 0. It always happens in F.binary_cross_entropy_with_logits which is used to compute q_halt_loss. I get nan during gradient computations eventually. 

 The code starts with `q_head.weight = 0` and `q_head.bias = -5.0`. The "ground truth" for binary_cross_entropy_with_logits  is set by `seq_is_correct.to(outputs["q_halt_logits"].dtype)`. This "ground truth" happens to be False or zero everywhere for all the steps during it = 0. This slowly drives the q_head weights and biases such that q_halt_logits goes towards negative infinity. I am observing this behavior using the following print statements in models/losses.py.

```
        print(f'outputs["q_halt_logits"]:min: {torch.min(outputs["q_halt_logits"])}')
        print(f'outputs["q_halt_logits"]:max: {torch.max(outputs["q_halt_logits"])}', flush=True)
```

I am wondering why others are not seeing this behavior. Is it possible for someone to share their wandb file(s)?

**Comment by bumi001** [comment]
@simplyrin2 , I am wondering if you observed NaNs using Nvidia GPUs? You can see train/lm_loss and train/q_halt_loss for arc-agi-1 dataset at https://huggingface.co/Sanjin2024/TinyRecursiveModels-ARC-AGI-1. In my case lm_loss comes down to around 1.85 similar to what is shown there and then gets stuck around that value. I am using the same code and configuration values shown there.
